### PR TITLE
Add support to insert a layer above a reference layer

### PR DESCRIPTION
**Pull request by JonasVautherin**
Right now, only the below layer is exposed. This exposes the above layer as well.

I do realise that there is some code generation setup in the repo (see [scripts](https://github.com/maplibre/maplibre-plugins-android/tree/main/plugin-annotation/scripts)), but I don't really know how to use them, or even if you are planning on keeping them (the repo does not seem very active to be honest).

Are the changes fine like this, or should I try to get them through the codegen? For the latter, how do you run the codegen?

@RomanBapst: FYI

### CODE REVIEW

**Comment by louwers** [comment]
@JonasVautherin Yes this repo is pretty stale. 

Please see this issue: https://github.com/maplibre/maplibre-native/issues/1189

We're planning to merge the Annotations functionality back into the main repo, for easier maintainability. If you could share your thoughts, also in the context of this issue, that would be very helpful.

**Comment by JonasVautherin** [comment]
> Yes this repo is pretty stale.

Does that mean we should just maintain and publish a fork until we see how it goes with #1189 and get a chance to have our changes merged there?

**Comment by fynngodau** [comment]
@JonasVautherin Definetly a good question! I was not aware that we had code generation in this repo at all.

To answer the question how to use it: provided that you are at the root of your repository, and that nodejs is installed,

1. Make sure the submodule is in place with `git submodule init && git submodule update`
2. Run `node plugin-annotation/scripts/code-gen.js`

I think we can extract multiple todos from this:

* document codegen
* sync codegen with current code (as we had multiple PRs that did not update the codegen code, including from me)

**Comment by JonasVautherin** [comment]
> sync codegen with current code (as we had multiple PRs that did not update the codegen code, including from me)

So now I think I have more questions :sweat_smile::

@louwers above seemed to suggest that those plugins would be rewritten soon. Should we bother getting the codegen in sync with the previous changes? I'm happy to make the codegen work for this PR, but it means that you will lose your previous changes, right?

**Comment by louwers** [comment]
@JonasVautherin Happy to merge this and make a release.

Yeah there is code gen involved, I haven't really looked into it much. 🥲

**Comment by louwers** [comment]
@RomanBapst Could you also review these changes?

@JonasVautherin Let me know when it is ready to merge.

**Comment by JonasVautherin** [comment]
> @RomanBapst Could you also review these changes?

Note that all I did was move the changes into the template and have the file auto-generated. So it's really the same thing, except for two `UiThread` annotations that we apparently did not need before (and that are now autogenerated).

> Let me know when it is ready to merge.

@louwers: I think it would be nice to first merge https://github.com/maplibre/maplibre-plugins-android/pull/36, such that we could see here that it passes the unit tests here :innocent:

File: plugin-annotation/src/main/java/com/mapbox/mapboxsdk/plugins/annotation/AnnotationManager.java
Comment (fynngodau) [inline plugin-annotation/src/main/java/com/mapbox/mapboxsdk/plugins/annotation/AnnotationManager.java:381]: The code appears to imply that `aboveLayerId` is ignored when `belowLayerId` is set. Do you think it would make sense to have a behavior noticeable to the developer to inform them that one parameter is being ignored?

In a Kotlinified version of this file (re. https://github.com/maplibre/maplibre-native/pull/1255) I would expect both of them to be optional parameters and an exception to be thrown if both are provided.

**Comment by fynngodau** [comment]
> I think it would be nice to first merge https://github.com/maplibre/maplibre-plugins-android/pull/36, such that we could see here that it passes the unit tests here

I merged #36, so if you were to rebase onto main, we would be able to get new test results :)

File: plugin-annotation/src/main/java/com/mapbox/mapboxsdk/plugins/annotation/AnnotationManager.java
Comment (JonasVautherin) [inline plugin-annotation/src/main/java/com/mapbox/mapboxsdk/plugins/annotation/AnnotationManager.java:379]: @fynngodau: Did you mean something like this?

Comment (fynngodau) [reply plugin-annotation/src/main/java/com/mapbox/mapboxsdk/plugins/annotation/AnnotationManager.java:379]: @JonasVautherin Yes, that's what I had in mind!

`IllegalArgumentException` could be a better fit? :thinking:

**Review by fynngodau — APPROVED** [review]
@JonasVautherin I agree :slightly_smiling_face:

@louwers I'll let you have a final look as well before merging.

**Pull request merged**
